### ISSUE TITLE
dark outline around links on cnet.com

### ISSUE THREAD

**Issue by nical** [servo/webrender#2595]
On this page: https://www.cnet.com/news/mozilla-open-source-firefox-move-helped-rewrite-tech-rules-anniversary/ the links have some sort of thin dark outline around them which looks like a sampling artifact. It was reproduced on linux and mac in the latest firefox nightly.

**Comment by glennw** [comment]
Here is a minimal repro:

```
---
root:
  items:
    - type: box-shadow
      bounds: [ 56.4833374023438, 50, 81, 25 ]
      color: black
      clip-mode: inset
      offset: 0 -4
```

I haven't confirmed, but I believe this will be an instance of this bug https://github.com/servo/webrender/issues/2501.

The bounds of the shadow are on a sub-pixel boundary, which is getting snapped to a device pixel. However, as per the bug above, the clip mask generation doesn't correctly deal with device snapping in this case.

Note to self: if this was being segmented, the problem would disappear - the segment builder logic eliminates the need for a clip mask, resulting in correct rendering. In this case we don't select segment building, since the area of the primitive is too small - we might want to revisit this heuristic, since in this case segmenting is probably worth it, to save a clip mask being generated. We still need to fix the clip snapping logic above though.

**Comment by glennw** [comment]
This will be fixed when #2608 lands.
